tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,1176 @@
+ *
+ * @a		- coefficient for y = a + b * x + eps.
+ * @b		- coefficient for y = a + b * x + eps.
+ * @x_avg	- average x value.
+ * @y_avg	- average y value.
+ * @xy_avg	- avg(x * y).
+ * @x_avg_y_avg	- avg(x) * avg(y).
+ * @x_sq_avg	- avg(x * x).
+ * @x_avg_sq	- avg(x) * avg(x).
+ */
+typedef struct {
+	long		a;
+	long		b;
+	long		x_avg;
+	long		y_avg;
+	long		xy_avg;
+	long		x_avg_y_avg;
+	long		x_sq_avg;
+	long		x_avg_sq;
+	TfwRatioHstXY	*hist;
+} TfwRatioHstDesc;
+
+/**
+ * Historic (past) data for predictive scheduler.
+ *
+ * @ahead	- predict for this number of @intvl ahead.
+ * @past_sz	- total number of slots for past data.
+ * @counter	- slot that is available for storing past data.
+ * @past	- past data for each server (@past[@srv_n]).
+ */
+typedef struct {
+	unsigned int	ahead;
+	size_t		past_sz;
+	unsigned long	counter;
+	TfwRatioHstDesc	*past;
+} TfwRatioHstData;
+
+/**
+ * The main Ratio Scheduler structure.
+ *
+ * All servers, either dead or live, are present in the list during
+ * the whole run-time. That may change in the future.
+ *
+ * @rcu		- RCU control structure;
+ * @free	- indicates that the pool entry is available for use.
+ * @srvdata	- scheduler data specific to each server in the group.
+ * @schdata	- scheduler data common to all servers in the group.
+ */
+typedef struct {
+	struct rcu_head		rcu;
+	atomic_t		free;
+	TfwRatioSrvData		*srvdata;
+	TfwRatioSchData		schdata;
+} TfwRatio;
+
+/**
+ * The pool of TfwRatio{} structures for RCU.
+ *
+ * @srv_n	- number of upstream servers.
+ * @psidx	- APM pstats[] value index for dynamic ratios.
+ * @intvl	- interval for re-arming the timer.
+ * @rpool	- pool of TfwRatio{} for RCU.
+ * @ratio	- pointer to the currently used structure.
+ * @hstdata	- historic data for predictive scheduler.
+ * @srvdesc	- array of upstream server descriptors.
+ * @rearm	- indicates if the timer can be re-armed.
+ * @timer	- periodic timer for dynamic APM data.
+ */
+typedef struct {
+	size_t			srv_n;
+	size_t			psidx;
+	unsigned int		intvl;
+	TfwRatio		*rpool;
+	TfwRatio __rcu		*ratio;
+	TfwRatioHstData		*hstdata;
+	TfwRatioSrvDesc		*srvdesc;
+	atomic_t		rearm;
+	struct timer_list	timer;
+} TfwRatioPool;
+
+/**
+ * Swap two server data entries. Required for sorting by sort().
+ */
+static void
+tfw_sched_ratio_srvdata_swap(void *lhs, void *rhs, int size)
+{
+	TfwRatioSrvData *lhs_data = (TfwRatioSrvData *)lhs;
+	TfwRatioSrvData *rhs_data = (TfwRatioSrvData *)rhs;
+	TfwRatioSrvData tmp = *lhs_data;
+	*lhs_data = *rhs_data;
+	*rhs_data = tmp;
+}
+
+/**
+ * Sort server data entries by ratio in descending order. Entries
+ * with higher ratios are moved towards the start of the array.
+ */
+static int
+tfw_sched_ratio_srvdata_cmp(const void *lhs, const void *rhs)
+{
+	unsigned int lhs_ratio = ((const TfwRatioSrvData *)lhs)->oratio;
+	unsigned int rhs_ratio = ((const TfwRatioSrvData *)rhs)->oratio;
+
+	if (lhs_ratio > rhs_ratio)
+		return -1;
+	if (lhs_ratio < rhs_ratio)
+		return 1;
+	return 0;
```

Comment:
If you need +/-1, you can use something like SGN on top of it

https://github.com/nmmmnu/HM3/blob/master/include/sgn.h

C++, but template can be removed easily. It will give you +1 / - 1